File: benchmarks/cpp/string concatenation and substring search/stringconcat_substringsearch.cpp
#include <iostream>
#include <string>
#include <chrono>

using namespace std;
using namespace chrono;

int main() {
    const int REPEAT = 1'000'0000; // Scale workload

    string chunk =
        "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789"
        "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789"
        "abcd"; // 100 chars

    string text;
    text.reserve(chunk.size() * REPEAT);

    string needle = "XYZ123";

    auto start = high_resolution_clock::now();

    // ---- String Concatenation ----
    for (int i = 0; i < REPEAT; i++) {
        text += chunk;
    }

    // ---- Substring Search ----
    size_t pos = text.find(needle);

    auto end = high_resolution_clock::now();
    duration<double> elapsed = end - start;

    cout << "C++ Concat + Search Time: "
         << elapsed.count() << " seconds\n";

    // Prevent over optimization
    cout << "Found at: " << pos << "\n";

    return 0;
}

File: benchmarks/rust/string concatenation and substring search/stringconcat_substringsearch.rs
use std::time::Instant;
use std::hint::black_box;

fn main() {
    let repeat: usize = 1_000_0000; // Scale workload

    let chunk =
        "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789\
         abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789\
         abcd"; // 100 chars

    let mut text = String::with_capacity(chunk.len() * repeat);
    let needle = "XYZ123";

    let start = Instant::now();

    // ---- String Concatenation ----
    for _ in 0..repeat {
        text.push_str(chunk);
    }

    // ---- Substring Search ----
    let pos = text.find(needle);

    let elapsed = start.elapsed();

    black_box(pos);

    println!(
        "Rust Concat + Search Time: {:.6} seconds",
        elapsed.as_secs_f64()
    );
}
